src/FsAutoComplete.Core/AbstractClassStubGenerator.fs

```diff
@@ -10,53 +10,67 @@ open FsToolkit.ErrorHandling
 
 
 type AbstractClassData =
-  | ObjExpr of baseTy: SynType * bindings: SynBinding list * overallRange: Range
-  | ExplicitImpl of baseTy: SynType * members: SynMemberDefn list * safeInsertPosition: Position
+  | ObjExpr of baseTy: SynType * bindings: SynBinding list * newExpression: Range * withKeyword: Range option
+
+  | ExplicitImpl of
+    baseTy: SynType *
+    members: SynMemberDefn list *
+    /// the place where the inherit expression is declared - the codefix should insert
+    /// at an indent of .Start.Column, but insert a newline after .End
+    inheritExpressionRange: Range
 
   member x.AbstractTypeIdentRange =
     match x with
-    | ObjExpr(baseTy, _, _)
-    | ExplicitImpl(baseTy, _, _) -> baseTy.Range
+    | ObjExpr(baseTy = baseTy)
+    | ExplicitImpl(baseTy = baseTy) -> baseTy.Range
 
   member x.TypeParameters =
     match x with
-    | ObjExpr(t, _, _)
-    | ExplicitImpl(t, _, _) -> expandTypeParameters t
+    | ObjExpr(baseTy = t)
+    | ExplicitImpl(baseTy = t) -> expandTypeParameters t
+
+let private (|ExplicitCtor|_|) =
+  function
+  | SynMemberDefn.Member(SynBinding(valData = SynValData(Some({ MemberKind = SynMemberKind.Constructor }), _, _)), _) ->
+    Some()
+  | _ -> None
 
 /// checks to see if a type definition inherits an abstract class, and if so collects the members defined at that
 let private walkTypeDefn (SynTypeDefn(info, repr, members, implicitCtor, range, trivia)) =
-  let reprMembers =
-    match repr with
-    | SynTypeDefnRepr.ObjectModel(_, members, _) -> members
-    | _ -> []
-
-  let allMembers = reprMembers @ (Option.toList implicitCtor) @ members
+  option {
+    let reprMembers =
+      match repr with
+      | SynTypeDefnRepr.ObjectModel(_, members, _) -> members // repr members already includes the implicit ctor if present
+      | _ -> Option.toList implicitCtor
+
+    let allMembers = reprMembers @ members
+
+    let! inheritType, inheritMemberRange = // this must exist for abstract types
+      allMembers
+      |> List.tryPick (function
+        | SynMemberDefn.ImplicitInherit(inheritType, inheritArgs, alias, range) -> Some(inheritType, range)
+        | _ -> None)
+
+    let furthestMemberToSkip, otherMembers =
+      ((inheritMemberRange, []), allMembers)
+      // find the last of the following kinds of members, as object-programming members must come after these
+      // * implicit/explicit constructors
+      // * `inherit` expressions
+      // * class-level `do`/`let` bindings (`do` bindings are actually `LetBindings` in the AST)
+      ||> List.fold (fun (m, otherMembers) memb ->
+        match memb with
+        | (SynMemberDefn.ImplicitCtor _ | ExplicitCtor | SynMemberDefn.ImplicitInherit _ | SynMemberDefn.LetBindings _) as possible ->
+          let c = Range.rangeOrder.Compare(m, possible.Range)
+
+          let m' = if c < 0 then possible.Range else m
+
+          m', otherMembers
+        | otherMember -> m, otherMember :: otherMembers)
```

Comment:
This is the core change - we no longer insert before-the-last member, instead we find the first available insert location that's valid and insert there. Easier to reason about IMO.